File: packages/mobile_agent_auth/ios/mobile_agent_auth/Sources/mobile_agent_auth/MobileAgentAuthPlugin.swift
import AppAuth
import Flutter
import Security
import UIKit

public final class MobileAgentAuthPlugin: NSObject, FlutterPlugin, FlutterApplicationLifeCycleDelegate, FlutterStreamHandler {
  private static let channelName = "mobile_agent_auth"
  private static let authEventChannelName = "mobile_agent_auth/events"
  private static let conversationChannelName = "mobile_agent_conversations"
  private static let protocolVersion = 1
  private static let conversationSchemaVersion = 1
  private let stateStore = KeychainAuthStateStore()
  private let conversationStore = SecureConversationStore()
  private let conversationQueue = DispatchQueue(label: "ai.coreline.mobileagent.conversations")
  private lazy var transport = NativeLlmTransportController(
    stateStore: stateStore,
    onReauthenticationRequired: { [weak self] in
      self?.emitAuthEvent(
        reason: "reauthentication_required",
        session: self?.reauthenticationRequired() ?? [:]
      )
    }
  )
  private var currentAuthorizationFlow: OIDExternalUserAgentSession?
  private var pendingResult: FlutterResult?
  private var authEventSink: FlutterEventSink?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: registrar.messenger())
    let instance = MobileAgentAuthPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
    registrar.addApplicationDelegate(instance)
    let authEvents = FlutterEventChannel(
      name: authEventChannelName,
      binaryMessenger: registrar.messenger()
    )
    authEvents.setStreamHandler(instance)
    let conversations = FlutterMethodChannel(
      name: conversationChannelName,
      binaryMessenger: registrar.messenger()
    )
    conversations.setMethodCallHandler(instance.handleConversation)
    let transportChannel = FlutterMethodChannel(
      name: "mobile_agent_llm_transport",
      binaryMessenger: registrar.messenger()
    )
    transportChannel.setMethodCallHandler(instance.transport.handle)
    let transportEvents = FlutterEventChannel(
      name: "mobile_agent_llm_transport/events",
      binaryMessenger: registrar.messenger()
    )
    transportEvents.setStreamHandler(instance.transport)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getCapabilities":
      result(capabilities())
    case "signIn":
      signIn(arguments: call.arguments as? [String: Any], result: result)
    case "restoreSession":
      restoreSession(result: result)
    case "cancelSignIn":
      cancelSignIn(result: result)
    case "signOut":
      signOut(arguments: call.arguments as? [String: Any], result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func handleConversation(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    conversationQueue.async { [weak self] in
      guard let self else { return }
      do {
        let value: Any?
        switch call.method {
        case "loadConversationSnapshot":
          let payload = try self.conversationStore.read()
          value = payload.map {
            [
              "schemaVersion": Self.conversationSchemaVersion,
              "payload": $0,
            ]
          }
        case "saveConversationSnapshot":
          guard let arguments = call.arguments as? [String: Any],
                Set(arguments.keys) == Set(["schemaVersion", "payload"]),
                arguments["schemaVersion"] as? Int == Self.conversationSchemaVersion,
                let payload = arguments["payload"] as? String
          else {
            throw ConversationVaultError.invalid
          }
          try self.conversationStore.write(payload)
          value = nil
        case "clearConversationSnapshot":
          guard call.arguments == nil else { throw ConversationVaultError.invalid }
          try self.conversationStore.clear()
          value = nil
        default:
          DispatchQueue.main.async { result(FlutterMethodNotImplemented) }
          return
        }
        DispatchQueue.main.async { result(value) }
      } catch ConversationVaultError.invalid {
        DispatchQueue.main.async {
          result(self.stableError("conversation_invalid", "Conversation payload is invalid."))
        }
      } catch {
        DispatchQueue.main.async {
          result(self.stableError("conversation_storage_failure", "Encrypted conversation storage failed."))
        }
      }
    }
  }

  private func signIn(arguments: [String: Any]?, result: @escaping FlutterResult) {
    guard pendingResult == nil else {
      result(stableError("operation_in_progress", "An authorization operation is active."))
      return
    }
    let configuration: OAuthConfiguration
    do {
      configuration = try OAuthConfiguration(arguments: arguments)
    } catch {
      result(stableError("configuration_invalid", "OIDC configuration is invalid."))
      return
    }
    guard let presenter = topViewController() else {
      result(stableError("activity_unavailable", "No foreground view controller is available."))
      return
    }

    pendingResult = result
    OIDAuthorizationService.discoverConfiguration(forIssuer: configuration.issuer) {
      [weak self] serviceConfiguration, _ in
      guard let self, self.pendingResult != nil else { return }
      guard let serviceConfiguration else {
        self.finishWithError("oauth_discovery_failed", "OIDC discovery failed.")
        return
      }
      var additionalParameters: [String: String]?
      if let audience = configuration.audience {
        additionalParameters = ["audience": audience]
      }
      let request = OIDAuthorizationRequest(
        configuration: serviceConfiguration,
        clientId: configuration.clientId,
        clientSecret: nil,
        scopes: configuration.scopes,
        redirectURL: configuration.redirectURI,
        responseType: OIDResponseTypeCode,
        additionalParameters: additionalParameters
      )
      self.currentAuthorizationFlow = OIDAuthState.authState(
        byPresenting: request,
        presenting: presenter
      ) { [weak self] authState, error in
        guard let self else { return }
        self.currentAuthorizationFlow = nil
        guard let callback = self.pendingResult else { return }
        self.pendingResult = nil
        guard let authState, authState.isAuthorized, error == nil else {
          let nsError = error as NSError?
          let cancelled = nsError?.code == -3 || nsError?.code == -4
          callback(self.stableError(
            cancelled ? "oauth_authorization_cancelled" : "oauth_authorization_failed",
            cancelled ? "Authorization was cancelled." : "Authorization did not complete."
          ))
          return
        }
        do {
          try self.stateStore.write(authState)
          self.transport.updateAuthState(authState)
          let summary = self.sessionSummary(authState)
          callback(summary)
          self.emitAuthEvent(reason: "signed_in", session: summary)
        } catch {
          self.stateStore.clear()
          callback(self.stableError("storage_failure", "Secure storage failed."))
        }
      }
    }
  }

  private func restoreSession(result: @escaping FlutterResult) {
    do {
      guard let state = try stateStore.read() else {
        let summary = signedOut()
        result(summary)
        emitAuthEvent(reason: "restored", session: summary)
        return
      }
      let summary: [String: Any]
      if state.isAuthorized {
        transport.updateAuthState(state)
        summary = sessionSummary(state)
      } else {
        summary = reauthenticationRequired()
      }
      result(summary)
      emitAuthEvent(reason: "restored", session: summary)
    } catch {
      stateStore.clear()
      result(stableError("storage_failure", "Secure storage failed."))
    }
  }

  private func cancelSignIn(result: @escaping FlutterResult) {
    currentAuthorizationFlow?.cancel()
    currentAuthorizationFlow = nil
    if let callback = pendingResult {
      pendingResult = nil
      callback(stableError("oauth_authorization_cancelled", "Authorization was cancelled."))
    }
    result(nil)
  }

  private func signOut(arguments: [String: Any]?, result: @escaping FlutterResult) {
    let allowedKeys = Set(["bffBaseUrl"])
    guard arguments == nil || Set(arguments!.keys).isSubset(of: allowedKeys) else {
      result(stableError("configuration_invalid", "Sign-out configuration is invalid."))
      return
    }
    let bffBaseURL: URL?
    if let rawBffBaseURL = arguments?["bffBaseUrl"] as? String {
      guard let validated = try? NativeStreamInput.validateBaseURL(rawBffBaseURL) else {
        result(stableError("configuration_invalid", "Sign-out configuration is invalid."))
        return
      }
      bffBaseURL = validated
    } else {
      bffBaseURL = nil
    }
    currentAuthorizationFlow?.cancel()
    currentAuthorizationFlow = nil
    if let callback = pendingResult {
      pendingResult = nil
      callback(stableError("oauth_authorization_cancelled", "Authorization was cancelled."))
    }
    transport.revokeSession(bffBaseURL: bffBaseURL) {
      result(nil)
      self.emitAuthEvent(reason: "signed_out", session: self.signedOut())
    }
  }

  private func finishWithError(_ code: String, _ message: String) {
    guard let callback = pendingResult else { return }
    pendingResult = nil
    callback(stableError(code, message))
  }

  private func sessionSummary(_ state: OIDAuthState) -> [String: Any] {
    var summary: [String: Any] = [
      "status": "authenticated",
      "canRefresh": state.refreshToken != nil,
    ]
    if let expiration = state.lastTokenResponse?.accessTokenExpirationDate {
      summary["expiresAtMillis"] = Int64(expiration.timeIntervalSince1970 * 1000)
    }
    let idToken = state.lastTokenResponse?.idToken ?? state.lastAuthorizationResponse.idToken
    if let accountLabel = accountLabel(from: idToken) {
      summary["accountLabel"] = accountLabel
    }
    return summary
  }

  private func accountLabel(from idToken: String?) -> String? {
    guard let payload = idToken?.split(separator: ".").dropFirst().first else { return nil }
    var base64 = String(payload).replacingOccurrences(of: "-", with: "+")
      .replacingOccurrences(of: "_", with: "/")
    base64 += String(repeating: "=", count: (4 - base64.count % 4) % 4)
    guard
      let data = Data(base64Encoded: base64),
      let claims = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    else { return nil }
    for key in ["name", "preferred_username", "email"] {
      if let value = claims[key] as? String, !value.isEmpty { return value }
    }
    return nil
  }

  private func signedOut() -> [String: Any] {
    ["status": "signed_out", "canRefresh": false]
  }

  private func reauthenticationRequired() -> [String: Any] {
    ["status": "reauthentication_required", "canRefresh": false]
  }

  private func capabilities() -> [String: Any] {
    [
      "protocolVersion": Self.protocolVersion,
      "authStateEvents": true,
      "secureSessionStorage": true,
      "nativeAuthorizedTransport": true,
    ]
  }

  private func emitAuthEvent(reason: String, session: [String: Any]) {
    let event: [String: Any] = [
      "protocolVersion": Self.protocolVersion,
      "reason": reason,
      "session": session,
    ]
    DispatchQueue.main.async { [weak self] in
      self?.authEventSink?(event)
    }
  }

  public func onListen(
    withArguments arguments: Any?,
    eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    authEventSink = events
    do {
      if let state = try stateStore.read() {
        let summary = state.isAuthorized
          ? sessionSummary(state)
          : reauthenticationRequired()
        emitAuthEvent(reason: "restored", session: summary)
      } else {
        emitAuthEvent(reason: "restored", session: signedOut())
      }
    } catch {
      stateStore.clear()
      emitAuthEvent(reason: "signed_out", session: signedOut())
    }
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    authEventSink = nil
    return nil
  }

  private func stableError(_ code: String, _ message: String) -> FlutterError {
    FlutterError(code: code, message: message, details: nil)
  }

  private func topViewController() -> UIViewController? {
    let scenes = UIApplication.shared.connectedScenes.compactMap { $0 as? UIWindowScene }
    let root = scenes.flatMap(\.windows).first(where: \.isKeyWindow)?.rootViewController
    var current = root
    while let presented = current?.presentedViewController { current = presented }
    return current
  }

  public func application(
    _ application: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey: Any] = [:]
  ) -> Bool {
    guard let flow = currentAuthorizationFlow else { return false }
    do {
      try flow.resumeExternalUserAgentFlow(with: url)
      currentAuthorizationFlow = nil
      return true
    } catch {
      return false
    }
  }
}

private struct OAuthConfiguration {
  let issuer: URL
  let clientId: String
  let redirectURI: URL
  let scopes: [String]
  let audience: String?

  init(arguments: [String: Any]?) throws {
    let allowedKeys = Set(["issuer", "clientId", "redirectUri", "scopes", "audience"])
    guard
      let arguments,
      Set(arguments.keys).isSubset(of: allowedKeys),
      let issuerString = arguments["issuer"] as? String,
      let issuer = URL(string: issuerString),
      issuer.scheme == "https",
      issuer.host != nil,
      URLComponents(url: issuer, resolvingAgainstBaseURL: false)?.query == nil,
      issuer.fragment == nil,
      let clientId = arguments["clientId"] as? String,
      clientId.range(of: "^[A-Za-z0-9._~-]{3,200}$", options: .regularExpression) != nil,
      let redirectString = arguments["redirectUri"] as? String,
      let redirectURI = URL(string: redirectString),
      let redirectComponents = URLComponents(url: redirectURI, resolvingAgainstBaseURL: false),
      Self.allowedRedirect(redirectString, redirectComponents),
      let scopes = arguments["scopes"] as? [String],
      scopes.contains("openid"),
      scopes.allSatisfy({
        $0.range(of: "^[A-Za-z0-9:._~-]+$", options: .regularExpression) != nil
      })
    else { throw OAuthConfigurationError.invalid }
    self.issuer = issuer
    self.clientId = clientId
    self.redirectURI = redirectURI
    self.scopes = scopes
    self.audience = (arguments["audience"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func allowedRedirect(
    _ rawValue: String,
    _ components: URLComponents
  ) -> Bool {
    if rawValue == "ai.coreline.mobileagent:/oauth/callback" { return true }
    return components.scheme == "https"
      && components.host != nil
      && components.user == nil
      && components.password == nil
      && components.query == nil
      && components.fragment == nil
  }
}

private enum OAuthConfigurationError: Error { case invalid }

final class KeychainAuthStateStore {
  private let service = "ai.coreline.mobileagent.oauth"
  private let account = "oidc-auth-state-v1"
  private let lock = NSRecursiveLock()

  func write(_ state: OIDAuthState) throws {
    lock.lock()
    defer { lock.unlock() }
    let data = try NSKeyedArchiver.archivedData(withRootObject: state, requiringSecureCoding: true)
    clear()
    let status = SecItemAdd([
      kSecClass: kSecClassGenericPassword,
      kSecAttrService: service,
      kSecAttrAccount: account,
      kSecAttrAccessible: kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
      kSecValueData: data,
    ] as CFDictionary, nil)
    guard status == errSecSuccess else { throw KeychainError.status(status) }
  }

  func read() throws -> OIDAuthState? {
    lock.lock()
    defer { lock.unlock() }
    var item: CFTypeRef?
    let status = SecItemCopyMatching([
      kSecClass: kSecClassGenericPassword,
      kSecAttrService: service,
      kSecAttrAccount: account,
      kSecReturnData: true,
      kSecMatchLimit: kSecMatchLimitOne,
    ] as CFDictionary, &item)
    if status == errSecItemNotFound { return nil }
    guard status == errSecSuccess, let data = item as? Data else {
      throw KeychainError.status(status)
    }
    return try NSKeyedUnarchiver.unarchivedObject(ofClass: OIDAuthState.self, from: data)
  }

  func clear() {
    lock.lock()
    defer { lock.unlock() }
    SecItemDelete([
      kSecClass: kSecClassGenericPassword,
      kSecAttrService: service,
      kSecAttrAccount: account,
    ] as CFDictionary)
  }
}

private enum KeychainError: Error { case status(OSStatus) }

File: packages/mobile_agent_auth/android/src/main/kotlin/ai/coreline/mobile_agent_auth/MobileAgentAuthPlugin.kt
package ai.coreline.mobile_agent_auth

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Base64
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.PluginRegistry
import java.nio.charset.StandardCharsets
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import net.openid.appauth.AuthState
import net.openid.appauth.AuthorizationException
import net.openid.appauth.AuthorizationRequest
import net.openid.appauth.AuthorizationResponse
import net.openid.appauth.AuthorizationService
import net.openid.appauth.AuthorizationServiceConfiguration
import net.openid.appauth.ResponseTypeValues
import org.json.JSONObject

class MobileAgentAuthPlugin :
    FlutterPlugin,
    MethodChannel.MethodCallHandler,
    EventChannel.StreamHandler,
    ActivityAware,
    PluginRegistry.ActivityResultListener {
    private lateinit var channel: MethodChannel
    private lateinit var transportChannel: MethodChannel
    private lateinit var transportEventChannel: io.flutter.plugin.common.EventChannel
    private lateinit var authEventChannel: EventChannel
    private lateinit var conversationChannel: MethodChannel
    private lateinit var applicationContext: Context
    private lateinit var stateStore: SecureAuthStateStore
    private lateinit var conversationStore: SecureConversationStore
    private var activity: Activity? = null
    private var authorizationService: AuthorizationService? = null
    private var transport: NativeLlmTransportController? = null
    private var pendingResult: MethodChannel.Result? = null
    private var authEventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())
    private val conversationExecutor: ExecutorService = Executors.newSingleThreadExecutor()

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        applicationContext = binding.applicationContext
        stateStore = SecureAuthStateStore(applicationContext)
        conversationStore = SecureConversationStore(applicationContext)
        authorizationService = AuthorizationService(applicationContext)
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
        authEventChannel = EventChannel(binding.binaryMessenger, AUTH_EVENT_CHANNEL)
        authEventChannel.setStreamHandler(this)
        conversationChannel = MethodChannel(binding.binaryMessenger, CONVERSATION_CHANNEL)
        conversationChannel.setMethodCallHandler(::onConversationMethodCall)
        transport = NativeLlmTransportController(
            stateStore,
            requireNotNull(authorizationService),
            onReauthenticationRequired = {
                emitAuthEvent("reauthentication_required", reauthenticationRequired())
            },
        )
        transportChannel = MethodChannel(binding.binaryMessenger, TRANSPORT_CHANNEL)
        transportChannel.setMethodCallHandler(transport)
        transportEventChannel = io.flutter.plugin.common.EventChannel(
            binding.binaryMessenger,
            TRANSPORT_EVENT_CHANNEL,
        )
        transportEventChannel.setStreamHandler(transport)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getCapabilities" -> result.success(capabilities())
            "signIn" -> signIn(call.arguments as? Map<*, *>, result)
            "restoreSession" -> restoreSession(result)
            "cancelSignIn" -> cancelSignIn(result)
            "signOut" -> signOut(call.arguments as? Map<*, *>, result)
            else -> result.notImplemented()
        }
    }

    private fun onConversationMethodCall(call: MethodCall, result: MethodChannel.Result) {
        conversationExecutor.execute {
            when (call.method) {
                "loadConversationSnapshot" -> completeConversation(result) {
                    conversationStore.read()?.let { payload ->
                        mapOf(
                            "schemaVersion" to CONVERSATION_SCHEMA_VERSION,
                            "payload" to payload,
                        )
                    }
                }
                "saveConversationSnapshot" -> completeConversation(result) {
                    val arguments = call.arguments as? Map<*, *> ?: error("arguments missing")
                    require(arguments.keys.all { it == "schemaVersion" || it == "payload" })
                    require(arguments["schemaVersion"] == CONVERSATION_SCHEMA_VERSION)
                    val payload = arguments["payload"] as? String ?: error("payload missing")
                    conversationStore.write(payload)
                    null
                }
                "clearConversationSnapshot" -> completeConversation(result) {
                    require(call.arguments == null)
                    conversationStore.clear()
                    null
                }
                else -> mainHandler.post(result::notImplemented)
            }
        }
    }

    private fun completeConversation(
        result: MethodChannel.Result,
        operation: () -> Any?,
    ) {
        try {
            val value = operation()
            mainHandler.post { result.success(value) }
        } catch (_: ConversationStoreException) {
            mainHandler.post {
                result.error(
                    "conversation_storage_failure",
                    "Encrypted conversation storage failed.",
                    null,
                )
            }
        } catch (_: Exception) {
            mainHandler.post {
                result.error(
                    "conversation_invalid",
                    "Conversation payload is invalid.",
                    null,
                )
            }
        }
    }

    private fun signIn(arguments: Map<*, *>?, result: MethodChannel.Result) {
        if (pendingResult != null) {
            result.error("operation_in_progress", "An authorization operation is active.", null)
            return
        }
        val currentActivity = activity
        if (currentActivity == null) {
            result.error("activity_unavailable", "No foreground Activity is attached.", null)
            return
        }
        val configuration = try {
            OAuthConfiguration.from(arguments)
        } catch (_: Exception) {
            result.error("configuration_invalid", "OIDC configuration is invalid.", null)
            return
        }

        pendingResult = result
        AuthorizationServiceConfiguration.fetchFromIssuer(Uri.parse(configuration.issuer)) {
                serviceConfiguration,
                error,
            ->
            if (pendingResult == null) return@fetchFromIssuer
            if (serviceConfiguration == null || error != null) {
                finishWithError(
                    "oauth_discovery_failed",
                    "OIDC discovery failed.",
                )
                return@fetchFromIssuer
            }
            val additional = buildMap {
                configuration.audience?.let { put("audience", it) }
            }
            val request = AuthorizationRequest.Builder(
                serviceConfiguration,
                configuration.clientId,
                ResponseTypeValues.CODE,
                Uri.parse(configuration.redirectUri),
            )
                .setScopes(configuration.scopes)
                .setAdditionalParameters(additional)
                .build()
            try {
                val intent = requireNotNull(authorizationService)
                    .getAuthorizationRequestIntent(request)
                currentActivity.startActivityForResult(intent, AUTH_REQUEST_CODE)
            } catch (_: Exception) {
                finishWithError("activity_unavailable", "Authorization UI could not start.")
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode != AUTH_REQUEST_CODE) return false
        val result = pendingResult ?: return true
        pendingResult = null
        if (data == null) {
            result.error(
                "oauth_authorization_cancelled",
                "Authorization was cancelled.",
                null,
            )
            return true
        }
        val response = AuthorizationResponse.fromIntent(data)
        val authorizationError = AuthorizationException.fromIntent(data)
        if (response == null) {
            val code = if (authorizationError?.error == "access_denied") {
                "oauth_authorization_failed"
            } else {
                "oauth_authorization_cancelled"
            }
            result.error(code, "Authorization did not complete.", null)
            return true
        }

        val authState = AuthState(response, authorizationError)
        requireNotNull(authorizationService).performTokenRequest(
            response.createTokenExchangeRequest(),
        ) { tokenResponse, tokenError ->
            authState.update(tokenResponse, tokenError)
            if (tokenResponse == null || tokenError != null || !authState.isAuthorized) {
                result.error(
                    "oauth_token_exchange_failed",
                    "Authorization code exchange failed.",
                    null,
                )
                return@performTokenRequest
            }
            try {
                stateStore.write(authState)
                transport?.updateAuthState(authState)
                val summary = sessionSummary(authState)
                result.success(summary)
                emitAuthEvent("signed_in", summary)
            } catch (_: SecureStoreException) {
                stateStore.clear()
                result.error("storage_failure", "Secure storage failed.", null)
            }
        }
        return true
    }

    private fun restoreSession(result: MethodChannel.Result) {
        val state = stateStore.read()
        val summary = when {
            state == null -> signedOut()
            state.isAuthorized -> sessionSummary(state)
            else -> reauthenticationRequired()
        }
        result.success(summary)
        emitAuthEvent("restored", summary)
    }

    private fun cancelSignIn(result: MethodChannel.Result) {
        val authorizationResult = pendingResult
        pendingResult = null
        activity?.finishActivity(AUTH_REQUEST_CODE)
        authorizationResult?.error(
            "oauth_authorization_cancelled",
            "Authorization was cancelled.",
            null,
        )
        result.success(null)
    }

    private fun signOut(arguments: Map<*, *>?, result: MethodChannel.Result) {
        if (arguments != null && arguments.keys.any { it != "bffBaseUrl" }) {
            result.error("configuration_invalid", "Sign-out configuration is invalid.", null)
            return
        }
        val bffBaseUrl = arguments?.get("bffBaseUrl") as? String
        if (bffBaseUrl != null && !NativeLlmTransportController.isValidBaseUrl(bffBaseUrl)) {
            result.error("configuration_invalid", "Sign-out configuration is invalid.", null)
            return
        }
        pendingResult?.error(
            "oauth_authorization_cancelled",
            "Authorization was cancelled.",
            null,
        )
        pendingResult = null
        val currentTransport = transport
        if (currentTransport == null) {
            stateStore.clear()
            result.success(null)
            emitAuthEvent("signed_out", signedOut())
            return
        }
        currentTransport.revokeSession(bffBaseUrl) {
            result.success(null)
            emitAuthEvent("signed_out", signedOut())
        }
    }

    private fun finishWithError(code: String, message: String) {
        val result = pendingResult ?: return
        pendingResult = null
        result.error(code, message, null)
    }

    private fun sessionSummary(state: AuthState): Map<String, Any> = buildMap {
        put("status", "authenticated")
        put("canRefresh", state.refreshToken != null)
        state.accessTokenExpirationTime?.let { put("expiresAtMillis", it) }
        accountLabel(state.idToken)?.let { put("accountLabel", it) }
    }

    private fun accountLabel(idToken: String?): String? {
        val payload = idToken?.split('.')?.getOrNull(1) ?: return null
        return runCatching {
            val decoded = Base64.decode(payload, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)
            val claims = JSONObject(String(decoded, StandardCharsets.UTF_8))
            listOf("name", "preferred_username", "email")
                .asSequence()
                .map(claims::optString)
                .firstOrNull(String::isNotBlank)
        }.getOrNull()
    }

    private fun signedOut(): Map<String, Any> = mapOf(
        "status" to "signed_out",
        "canRefresh" to false,
    )

    private fun reauthenticationRequired(): Map<String, Any> = mapOf(
        "status" to "reauthentication_required",
        "canRefresh" to false,
    )

    private fun capabilities(): Map<String, Any> = mapOf(
        "protocolVersion" to PROTOCOL_VERSION,
        "authStateEvents" to true,
        "secureSessionStorage" to true,
        "nativeAuthorizedTransport" to true,
    )

    private fun emitAuthEvent(reason: String, summary: Map<String, Any>) {
        val event = mapOf(
            "protocolVersion" to PROTOCOL_VERSION,
            "reason" to reason,
            "session" to summary,
        )
        mainHandler.post { authEventSink?.success(event) }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        authEventSink = events
        val state = stateStore.read()
        val summary = when {
            state == null -> signedOut()
            state.isAuthorized -> sessionSummary(state)
            else -> reauthenticationRequired()
        }
        emitAuthEvent("restored", summary)
    }

    override fun onCancel(arguments: Any?) {
        authEventSink = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        activity = null
        finishWithError("activity_unavailable", "Foreground Activity was detached.")
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        authEventChannel.setStreamHandler(null)
        conversationChannel.setMethodCallHandler(null)
        transportChannel.setMethodCallHandler(null)
        transportEventChannel.setStreamHandler(null)
        transport?.dispose()
        transport = null
        authorizationService?.dispose()
        authorizationService = null
        pendingResult = null
        authEventSink = null
        conversationExecutor.shutdownNow()
    }

    companion object {
        private const val CHANNEL = "mobile_agent_auth"
        private const val AUTH_EVENT_CHANNEL = "mobile_agent_auth/events"
        private const val CONVERSATION_CHANNEL = "mobile_agent_conversations"
        private const val TRANSPORT_CHANNEL = "mobile_agent_llm_transport"
        private const val TRANSPORT_EVENT_CHANNEL = "mobile_agent_llm_transport/events"
        private const val AUTH_REQUEST_CODE = 0x4D41
        private const val PROTOCOL_VERSION = 1
        private const val CONVERSATION_SCHEMA_VERSION = 1
    }
}
